### PR TITLE
Renovation of the zenstore tests

### PR DESCRIPTION
**Pull request by AlexejPenner**
## Describe changes
I renovated the zenstore tests so that they will work with whatever zen_store is used by the test harness (sql_zen_store with sqlite backend, sql_zen_store with mysql backend, rest_zen_store)

As a first step i separated out the basic crud test that are the same for all entities from the entity specific tests. 

As a next step I would imagine tests, that test behaviour that might be different based on the backend (Think stack sharing)

Stacks and Step_Runs are currently not included in the basic crud tests, as both of these have dependencies on other entities being there and the abstraction currently does not capture that yet.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Review by fa9r — CHANGES_REQUESTED** [review]
Great design, love the CRUD test abstraction in particular!

Overall the implementation also looks good, but I think a few tests got lost in the transition:
- CRUD tests for step runs
- CRUD tests for schedules (perhaps not merged into this branch yet?)
- `test_getting_team_for_user`
- `test_access_user_in_team_succeeds`
- `test_roles_property_with_fresh_store`
- `test_list_stack_components_works_with_filters`
- `test_list_stack_components_lists_nothing_for_nonexistent_filters`

The last two in particular made me think whether we should also include some tests for filtering for the entities we iterate over. I think you assume that all these entities have names, in that case, we could add another test that checks that filtering by name works as expected (perhaps even using different filter options like `startswith`, `endswith`, ...).

Also, short warning already, you'll get quite a few merge conflicts from #1151, let me know when you want to merge that in so I can help with resolving them.

File: .pyspelling-ignore-words [hunk 1]
```diff
@@ -692,6 +695,11 @@ Standarize
 StepContext
 StepContextError
 StepDecorator
+StepRunFilterModel
+StepRunUpdateModel
+StepConfiguration
+StepRunRequestModel
+StepSpec
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply .pyspelling-ignore-words:709]: Thanks for that :+1:

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 2]
```diff
@@ -0,0 +1,991 @@
...
+    if isinstance(create_model, WorkspaceScopedRequestModel):
+        create_model.user = client.active_user.id
+        create_model.workspace = client.active_workspace.id
+    # Test the creation
+    created_entity = crud_test_config.create_method(create_model)
+    # Filter by name to verify the entity was actually created
+    entities_list = crud_test_config.list_method(
+        crud_test_config.filter_model(name=create_model.name)
+    )
+    assert entities_list.total > 0
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:96]:
```suggestion
    assert entities_list.total == 1
```

Because you only create one, right ? And the name is random, so even if you run it in parallel with other tests on the same or on other machines, it will work.

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 3]
```diff
@@ -0,0 +1,991 @@
...
+    # Filter by id to verify the entity was actually created
+    entities_list = crud_test_config.list_method(
+        crud_test_config.filter_model(id=created_entity.id)
+    )
+    assert entities_list.total > 0
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:101]:
```suggestion
    assert entities_list.total == 1
```

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 4]
```diff
@@ -0,0 +1,991 @@
...
+    DEFAULT_STACK_NAME,
+    DEFAULT_USERNAME,
+    DEFAULT_WORKSPACE_NAME,
+)
+
+DEFAULT_NAME = "default"
+
+# .--------------.
+# | GENERIC CRUD |
+# '--------------'
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:73]: This is the best thing I've seen lately !

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 5]
```diff
@@ -0,0 +1,991 @@
...
+                team_id=new_team.id, team_update=team_update
+            )
+
+            assert created_user.id in team_update.user_ids
+            assert len(team_update.users) == 1
+
+            # Make sure the team name has not been inadvertently changed
+            assert zen_store.get_team(new_team.id).name == team_name
+        # Cleanup no matter what
+        finally:
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:274]: Isn't this what the `TeamContext`  you created is for ?

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 6]
```diff
@@ -0,0 +1,991 @@
...
+
+        nonexistent_id = uuid.uuid4()
+
+        team_update = TeamUpdateModel(users=[nonexistent_id])
+        with pytest.raises(KeyError):
+            team_update = zen_store.update_team(
+                team_id=new_team.id, team_update=team_update
+            )
+    # Cleanup no matter what
+    finally:
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:299]: Same here: can you use `TeamContext` to do the cleanup ?

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 7]
```diff
@@ -0,0 +1,991 @@
...
+
+
+def test_removing_user_from_team_succeeds():
+    """Tests removing a user from a team."""
+
+    zen_store = Client().zen_store
+    team_name = sample_name("arias_team")
+
+    with UserContext() as created_user:
+        try:
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:315]: TeamContext to the rescue !

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 8]
```diff
@@ -0,0 +1,991 @@
...
+    with does_not_raise():
+        (zen_store.create_user_role_assignment(role_assignment))
+    with pytest.raises(IllegalOperationError):
+        zen_store.delete_role(created_role.id)
+
+    # Cleanup
+    with does_not_raise():
+        # By deleting the user first, the role assignment is cleaned up as well
+        zen_store.delete_user(created_user.id)
+        zen_store.delete_role(created_role.id)
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:444]: Could you use UserContext and RoleContext here instead ?

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 9]
```diff
@@ -0,0 +1,991 @@
...
+    zen_store = Client().zen_store
+    new_role = RoleRequestModel(name=sample_name("cat"), permissions=set())
+    created_role = zen_store.create_role(new_role)
+    with does_not_raise():
+        zen_store.get_role(role_name_or_id=new_role.name)
+    list_of_roles = zen_store.list_roles(RoleFilterModel(name=new_role.name))
+    assert list_of_roles.total > 0
+    # Cleanup
+    with does_not_raise():
+        zen_store.delete_role(created_role.id)
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:391]: RoleContext should help with this, but maybe you're testing a corner case where this actually doesn't make sense. In that case, maybe a try/catch ?

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 10]
```diff
@@ -0,0 +1,991 @@
...
+# | ROLE ASSIGNMENTS |
+# '------------------'
+
+
+def test_assigning_role_to_user_succeeds():
+    """Tests assigning a role to a user."""
+    zen_store = Client().zen_store
+
+    with RoleContext() as created_role:
+        with UserContext() as created_user:
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:457]: Beautiful ! 🚀

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 11]
```diff
@@ -0,0 +1,991 @@
...
+        c_type=StackComponentType.ORCHESTRATOR, flavor="local", config={}
+    ) as orchestrator:
+        with ComponentContext(
+            c_type=StackComponentType.ARTIFACT_STORE, flavor="local", config={}
+        ) as artifact_store:
+            components = {
+                StackComponentType.ORCHESTRATOR: [orchestrator.id],
+                StackComponentType.ARTIFACT_STORE: [artifact_store.id],
+            }
+            with StackContext(components=components) as stack:
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:696]: Wow ! You have unlocked Contextception !

I was about to suggest [contextlib.nested](https://docs.python.org/2/library/contextlib.html#contextlib.nested) to help with nested contexts, but now I see they depend on eachother (the stack does at least).

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 12]
```diff
@@ -0,0 +1,991 @@
...
+        # TODO: Implement this
+        assert True
+
+
+def test_updating_default_user_fails():
+    """Tests that updating the default user is prohibited."""
+    client = Client()
+    default_user = client.zen_store.get_user(DEFAULT_USERNAME)
+    assert default_user
+    user_update = UserUpdateModel(name="axl")
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:361]: Axl approves 😸

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 13]
```diff
@@ -0,0 +1,991 @@
...
+                stack = store.get_stack(created_stack.id)
+                assert stack is not None
+
+            # Update
+            stack_update = StackUpdateModel(name="axls_stack")
+            store.update_stack(stack_id=stack.id, stack_update=stack_update)
+
+            stacks = store.list_stacks(StackFilterModel(name="axls_stack"))
+            assert len(stacks) == 1
+            stacks = store.list_stacks(StackFilterModel(name="arias_stack"))
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:739]:
```suggestion
            stack_name = sample_name("arias_stack")
            new_stack = StackRequestModel(
                name=stack_name,
                components=components,
                workspace=client.active_workspace.id,
                user=client.active_user.id,
            )
            created_stack = store.create_stack(stack=new_stack)

            stacks = store.list_stacks(StackFilterModel(name=stack_name))
            assert len(stacks) == 1

            with does_not_raise():
                stack = store.get_stack(created_stack.id)
                assert stack is not None

            # Update
            stack_update = StackUpdateModel(name="axls_stack")
            store.update_stack(stack_id=stack.id, stack_update=stack_update)

            stacks = store.list_stacks(StackFilterModel(name="axls_stack"))
            assert len(stacks) == 1
            stacks = store.list_stacks(StackFilterModel(name=stack_name))
```

Should at least make the test parallelizable.

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 14]
```diff
@@ -0,0 +1,991 @@
...
+        c_type=StackComponentType.ORCHESTRATOR, flavor="local", config={}
+    ) as orchestrator:
+        with ComponentContext(
+            c_type=StackComponentType.ARTIFACT_STORE, flavor="local", config={}
+        ) as artifact_store:
+            components = {
+                StackComponentType.ORCHESTRATOR: [orchestrator.id],
+                StackComponentType.ARTIFACT_STORE: [artifact_store.id],
+            }
+            new_stack = StackRequestModel(
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:718]: StackContext should help here, no ?

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 15]
```diff
@@ -0,0 +1,991 @@
...
+
+
+def test_deleting_run_deletes_steps():
+    """Tests deleting run deletes its steps."""
+    client = Client()
+    store = client.zen_store
+
+    # Just in case the test db is not in a clean state we compare relative
+    num_steps_before = store.list_run_steps(StepRunFilterModel()).total
+    num_pipelines_before = store.list_runs(PipelineRunFilterModel()).total
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:901]: This will fail intermittently when run in parallel with other tests.

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 16]
```diff
@@ -0,0 +1,991 @@
...
+def test_list_unused_artifacts():
+    """Tests listing with `unused=True` only returns unused artifacts."""
+    client = Client()
+    store = client.zen_store
+
+    num_artifacts_before = store.list_artifacts(ArtifactFilterModel()).total
+    num_unused_artifacts_before = store.list_artifacts(
+        ArtifactFilterModel(only_unused=True)
+    ).total
+    num_runs = 1
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:963]: Same here, this will fail if run in parallel (local machine or different machines)

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 17]
```diff
@@ -0,0 +1,991 @@
...
+        artifacts = store.list_artifacts(ArtifactFilterModel(only_unused=True))
+        assert artifacts.total == num_unused_artifacts_before
+
+
+def test_artifacts_are_not_deleted_with_run():
+    """Tests listing with `unused=True` only returns unused artifacts."""
+    client = Client()
+    store = client.zen_store
+
+    num_artifacts_before = store.list_artifacts(ArtifactFilterModel()).total
```
Comment (stefannica) [inline tests/integration/functional/zen_stores/test_zen_store.py:978]: Same thing here.

**Review by stefannica — CHANGES_REQUESTED** [review]
Beautifully crafted ! I had hoped we'd be able to run these tests in parallel now, but there are still some minor details that stand in the way of that majestic goal. You can try this yourself with:

```
./zen-test deployment up default
pytest tests/integration/functional/zen_stores --environment default --no-provision -n 4
```

For one, the pipelines seem to be failing because "RuntimeError: Main module was not started from a file. Cannot determine the module root path", which I can't fathom how we could solve.

File: tests/integration/functional/zen_stores/test_zen_store.py (hunk 5 above)
Comment (AlexejPenner) [reply tests/integration/functional/zen_stores/test_zen_store.py:274]: Jup, the Conteyxts came halfway through thsi endevour, so might have missed some opportunities to backport

File: tests/integration/functional/zen_stores/test_zen_store.py (hunk 14 above)
Comment (AlexejPenner) [reply tests/integration/functional/zen_stores/test_zen_store.py:718]: Here we are explicitely testing the crud on stacks, I feel the actual logic being tested should not be abstracted away in a context

**Review by stefannica — APPROVED** [review]
I hopefully fixed the bugs affecting the remaining failing tests. If the CI passes, this is good to merge.

**Comment by stefannica** [comment]
> I hopefully fixed the bugs affecting the remaining failing tests. If the CI passes, this is good to merge.

I guess I didn't. There was one other issue that caused an outdated global configuration to be written to disk and picked up on the next run. Should be fixed now.
